test_workspace/recordset.bosatsu

```diff
@@ -1,37 +1,83 @@
-package RecordSet
-
-enum RecordValue:
-  RVBool(value: Bool)
-  RVInt(value: Int)
-  RVString(value: String)
-struct Field[t](name: String, toValue: t -> RecordValue)
-struct RecordSet[shape](fields: shape[Field[String], Field[Int], Field[Bool]], rows: List[shape[String,Int, Bool]])
-
-def restructure(RecordSet(fields, rows), f: forall string,int,bool. shape1[string, int, bool] -> shape2[string, int, bool]) -> RecordSet[shape2]:
-  RecordSet(f(fields), rows.map_List(f))
+package RecordSet/Library
+
+enum RowEntry[w]:
+  REBool(value: w[Bool])
+  REInt(value: w[Int])
+  REString(value: w[String])
+
+struct RecordField[t](name: String, to_entry: forall w. w[t] -> RowEntry[w])
+struct RecordValue[t](value: t)
+struct RecordGetter[shape, t](
+  field: shape[RecordField] -> RecordField[t],
+  value: shape[RecordValue] -> RecordValue[t]
+)
+struct RecordRowEntry[w, t](row_entry: RowEntry[w])
+
+struct RecordSet[shape](
+  fields: shape[RecordField],
+  rows: List[shape[RecordValue]],
+  getters: shape[RecordGetter[shape]],
+  traverse: forall w1,w2. shape[w1] -> (forall ss. w1[ss] -> w2[ss]) -> shape[w2],
+  record_to_list: forall w. shape[RecordRowEntry[w]] -> List[RowEntry[w]]
+)
 
-struct IntShape[string,int,bool](int: int)
-struct PairShape[shape,t,string,int,bool](left: t[string,int,bool], right: shape[string, int, bool])
-def make_pair_int_shape(rest: shape[string,int,bool], i: int) -> PairShape[shape,IntShape,string,int,bool]:
-  PairShape(IntShape(i), rest)
+def get(sh: shape[RecordValue], RecordGetter(_, getter): RecordGetter[shape, t]) -> t:
+  RecordValue(result) = sh.getter
+  result
+
+def create_field(rf: RecordField[t], fn: shape[RecordValue] -> t):
+  RecordGetter(\sh -> rf, \sh -> RecordValue(fn(sh)))
+
+def list_of_rows(RecordSet(fields, rows, getters, traverse, record_to_list): RecordSet[shape]):
+  def getter_to_row_entry(row: shape[RecordValue]):
+    (result_fn: forall tt. RecordGetter[shape, tt] -> RecordRowEntry[RecordValue, tt]) = \RecordGetter(get_field, get_value) ->
+      RecordField(name, to_entry) = get_field(fields)
+      RecordRowEntry(to_entry(get_value(row)))
+    result_fn
+  rows.map_List(\row -> record_to_list(getters.traverse(getter_to_row_entry(row))))
```

Comment:
you might consider `[record_to_list(getters.traverse(getter_to_row_entry(row))) for row in rows]` this will compile to the same code you have here but may be a bit prettier.... up to you.

Reply:
this change introduces the error
```
failure to get type: TyApply(TyVar(Bound(a)),TyConst(Defined(PackageName(NonEmptyList(RecordSet, Library)),TypeName(Constructor(RecordValue)))))
```

interestingly it only occurs when I run the tests and not when I typecheck the file

Reply:
That error message happens here:

https://github.com/johnynek/bosatsu/blob/master/core/src/main/scala/org/bykn/bosatsu/Evaluation.scala#L346

That looks like a bug of some kind. Testing uses evaluation, typechecking does not, and this bug looks like it happens with evaluation.